platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1354,19 +1378,36 @@ private void validateParentMutation(MutationType type) {
         }
 
         preventIllegalParentMutation(type);
-        markAsModified(type);
         notifyChildren(type);
+        maybePreventMutation(type, type + " of parent");
     }
 
     @Override
     public void validateMutation(MutationType type) {
         preventIllegalMutation(type);
-        markAsModified(type);
         notifyChildren(type);
+        maybePreventMutation(type, type.toString());
+    }
+
+    /**
+     * Emit a warning (and eventually throw an exception) if a mutation of type {@code type} occurs
+     * during a forbidden state.
+     */
+    private void maybePreventMutation(MutationType type, String typeDescription) {
+        // If an external party has seen the public state (variant metadata) of our configuration,
+        // we forbid any mutation that mutates the public state. The resolution strategy does
+        // not mutate the public state of the configuration, so we allow it.
+        if (observed && type != MutationType.STRATEGY) {
+            DeprecationLogger.deprecateBehaviour(String.format("Mutating the %s of %s after it has been resolved or consumed.", typeDescription, this.getDisplayName()))
+                .withAdvice("After a Configuration has been resolved, observed via dependency-management, or published, it should not be modified further.")
+                .willBecomeAnErrorInGradle9()
+                .withUpgradeGuideSection(8, "mutate_configuration_after_locking")
```

Comment:
❌ Where is this anchor supposed to go? It looks broken and it's not added in this PR.